### DISCUSSION TITLE
Proposal: primary constructors on records allow for object initializer syntax

### DISCUSSION THREAD

**Discussion by chase-miller** [dotnet/csharplang#8052] [Language Ideas]
### Summary
Update c# so that all records declared via primary constructor syntax can be instantiated either via constructor syntax or object initializer syntax. All constructor parameters in the record declaration function as `required` properties such that the consumer must provide them when opting for object initializer syntax. 

### Example
Record declaration:
```cs
public record MyData(string TenantId, string MyCoolVal)
{
    public string[] SomeOptionalProperty { get; init; } = Array.Empty<string>();
}
```

Initialization via object initializer syntax:
```cs
return new MyData
{
    TenantId = "SomeTenant",
    MyCoolVal = "SomeVal",
};
```

-- or --

Initialization via constructor:
```cs
return new MyData(
    TenantId: "SomeTenant",
    MyCoolVal: "SomeVal");
```

This would fail to compile because `TenantId` is a required property:
```cs
return new MyData
{
    MyCoolVal = "SomeVal",
};
```

### Workaround
We can achieve roughly equivalent functionality from the consumer's perspective via a much more verbose record declaration:
```cs
public record MyData
{
    public MyData()
    { }

    [SetsRequiredMembers]
    public MyData(string TenantId, string MyCoolVal)
    {
        this.TenantId = TenantId;
        this.MyCoolVal = MyCoolVal;
    }

    public void Deconstruct(out string tenantId, out string myCoolVal)
    {
        tenantId = TenantId;
        myCoolVal = MyCoolVal;
    }

    public required string TenantId { get; init; }
    public required string MyCoolVal { get; init; }

    public string[] SomeOptionalProperty { get; init; } = Array.Empty<string>();
}
```

### Background and Motivation
Facilitate terser and more readable record declarations for instances where we want to give a consumer the choice of object initializer or constructor syntax.

**Comment by tacosontitan** [comment]
> Facilitate terser and more readable record declarations for instances where we want to give a consumer the choice of object initializer or constructor syntax.

Terse does not always improve legibility. In this case, your proposal is violating expectations on behavior for constructors and object initializer syntax. A constructor specifies *required* data while object initializer syntax allows for further specification of the instance at initialization and only members explicitly marked as `required` but not set by the constructor are *required* to be set at that time.

You can see that with this sample snippet:
```cs
// no error about fizz not being set.
var sample = new Sample(foo: "foo") { Bar = "bar" };

// error because bar is not set but is required
var sample = new Sample(foo: "foo") { Fizz = 3 };

public sealed class Sample(string foo)
{
    public string Foo { get; set; } = foo;
    public required string Bar { get; set; }
    public int Fizz { get; init; }
}
```

**Reply by chase-miller** [reply]
Can you help me understand the semantic difference between these two snippets of code?

```cs
public record Sample(string Foo);
```

```cs
public record Sample
{
    public required string Foo { get; init; }
}
```

As I understand it they both convey to the consumer that `Foo` must be provided upon initialization of `Sample`.

**Reply by tacosontitan** [reply]
They *do* convey that `Foo` must be provided, but *how* and *when* they are required to be provided is different.

My point isn't that the two aren't synonymous, it's that your proposal recommends allowing them to be interchangeable which violates expectations. A constructor parameter must be set through the constructor and a required member must be set through object initialization syntax; so imagine specifying a constructor parameter which sets `Foo` but also marking `Foo` as required explicitly:

```cs
public record Sample(string foo)
{
    public required string Foo { get; set; } = foo;
}
```

This situation is bad because the argument passed to the constructor is immediately overwritten by the fact that `Foo` must be set using object initialization syntax which occurs after the constructor is invoked:

```cs
var sample = new Sample(foo: "foo") { Foo = "bar" };
```

<sub>*You can see this for yourself on [sharplab.io](https://sharplab.io/#v2:C4LgTgrgdgPgAgJgIwFgBQ6D6ACAvNqAUwHdsBlAQwFsAHAG0IAoAzAe1ZGwCI3WuBKbAG9sAMXZ5uAIwpgu2AL4BudHCQBORlwAirIgDoBKjGjgBmbGEIBjVmAAm5avULoh6bJ+w0wASwBuFMCE2GoADNiYvMZeoRZWAI4QvlaO4WLsHl7uaLGxAOaEwHgAfJHRWXnYAM5FlXk5VVVqmgAkXGRFwL5Q+di82MCs2AAGQoF0EIQKI4b8MU2xURL4E1MLVQr1W7le9eZOtAyM6byCuCX1seLD+BVoCkA=).*</sub>

The important take away is that the constructor is invoked *before* object initialization occurs, and having the two processes be considered interchangeable violates established expectations of that behavior.

**Reply by chase-miller** [reply]
Ah that resonates. Thanks for taking the time to help me understand; this is something I hadn't considered. Were this idea to be entertained further perhaps there must be a rule of mutual exclusivity when instantiating an object: either set via constructor or initializer.

That said, don't primary constructors already exhibit the behavior you mentioned whereby a property can be set via both the constructor _and_ property on initialization? Example:

```cs
public record Sample(string Foo);

[Test]
public void InitializerStompsOnConstructor()
{
    var sample = new Sample("foo")
    {
        Foo = "bar",
    };
   
    // passes
    sample.Foo.Should().Be("bar"); 
}
```

If consensus continues to grow that this isn't a great idea, maybe what I'm really looking for is a less verbose way to declare required properties (plus getting the benefit of other primary constructor perks like a compiler-created `Deconstruct` method). An additional consideration, however, might be how to do this in a way that also achieves the benefit of simpler object hierarchy as discussed [here](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-11.0/required-members#motivation): is there a way for a record to extend a base record via the proposed syntax without needing to redefine all the base record's required properties?

Sorry for being dense, and thanks for taking the time to discuss :-)

**Reply by tacosontitan** [reply]
> That said, don't primary constructors already exhibit the behavior you mentioned whereby a property can be set via both the constructor and property on initialization?

Absolutely, but at that point the constructor argument is still *required* and the developer is simply shooting themselves in the foot.

**Comment by HaloFour** [comment]
You could use a source generator to emit a partial type for the record containing a parameterless constructor.  But it's intentional that a positional record is required to be created through the primary constructor.
